config/config.go

```diff
@@ -0,0 +1,176 @@
+package config
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/pflag"
+	c "nav/constants"
+)
+
+type Config struct {
+	ConfValues ConfValues
+	// add DB connection instance here
+}
+
+type ConfValues struct {
+	Symbol         string    `json:"symbol"`
+	Type           string    `json:"output_type"`
+	DBDriver       string    `json:"db_driver"`
+	DBDSN          string    `json:"DBDSN"`
+	ExcludedBefore []string  `json:"excluded_before"`
+	ExcludedAfter  []string  `json:"excluded_after"`
+	TargetSubsys   []string  `json:"target_subsys"`
+	MaxDepth       int       `json:"max_depth"`
+	Mode           c.OutMode `json:"mode"`
+	DBInstance     int       `json:"db_instance"`
+}
+
+// New creates a new Config instance and returns a pointer to it.
+func New() (*Config, error) {
+	config := Config{}
+	confValues, err := initConfig()
+	config.ConfValues = confValues
+	if err != nil {
+		return nil, err
+	}
+	return &config, nil
+}
+
+// initConfig gathers the configuration values from command line and config file (if provided) and returns
+// them in a ConfValues struct. If both are provided, the command line flags take precedence.
+func initConfig() (ConfValues, error) {
+	fs := pflag.NewFlagSet("nav", pflag.ContinueOnError)
+	var confValues ConfValues
+	var configPath string
+
+	err := parseCommandLine(fs, &configPath, os.Args)
+	if err != nil {
+		return ConfValues{}, fmt.Errorf("error initializing command line flags: %w", err)
+	}
+
+	if len(os.Args) == 1 {
+		return ConfValues{}, fmt.Errorf("error: no configuration was specified, please specify a configuration file or use the command line flags; " +
+			"use -h or --help for more information")
+	}
+
+	if configPath != "" {
+		err = loadConfigFile(&confValues, configPath)
+		if err != nil {
+			return ConfValues{}, fmt.Errorf("error: %w", err)
+		}
+	}
+
+	setFlags(fs, &confValues)
+	if err := confValues.validate(); err != nil {
+		return ConfValues{}, fmt.Errorf("invalid configuration: %w", err)
+	}
+
+	return confValues, nil
+}
+
+func loadConfigFile(cfg *ConfValues, configPath string) error {
+	cleanedPath := filepath.Clean(configPath)
+
+	if cleanedPath != configPath {
```

Comment:
This seems unnecessarily strict to me. Why disallow "unclean" paths? Maybe they are automatically constructed and contain /./ or other redundancies, but why not?

Reply:
I think it's always good to validate and sanitize user input. Maybe in the future it will turn out that this kind of sanitization causes more problems than good and will need to be changed, but I think it's better than having none at all.

Reply:
I see. Especially when we start feeding parameters from the web it makes sense to be careful. But then I suggest we give the user more information about the issue, because I would not expect to be an unclean path to be invalid and it would take me time to find the problem. Example (makes only sense if some primitive script generated the path, of course, nobody would type that):
$ ./nav -f ../nav/../nav//example_config.json 
error: invalid file path

nicer:
$ ./nav -f ../nav/../nav//example_config.json 
error: invalid file path, did you mean '../nav/example_config.json'?

Reply:
Addressing in my next commit, marking as resolved